src/backend/mpi/write_buffer.hpp

```diff
@@ -227,83 +418,115 @@ class write_buffer
 		 * @param	val The value of type T to erase
 		 */
 		void erase(T val) {
-			std::lock_guard<std::mutex> lock(_buffer_mutex);
+			double t_start = MPI_Wtime();
+			// Delegate erasing to lock holder (can be self)
+			_qd_lock.DELEGATE_N(&write_buffer::_erase_helper, this, val);
+			double t_end = MPI_Wtime();
 
-			typename std::deque<T>::iterator it = std::find(_buffer.begin(),
-					_buffer.end(), val);
-			if(it != _buffer.end()){
-				_buffer.erase(it);
-			}
+			std::lock_guard<std::mutex> stat_lock(_stat_mutex);
+			_buffer_lock_time += t_end - t_start;
 		}
 
 		/**
 		 * @brief	Flushes the ArgoDSM write buffer to memory
 		 * @pre		Require ibsem to be taken until parallel MPI
 		 */
 		void flush() {
+			// Use an atomic flag to detect when flush is done
+			std::atomic<bool> w_flag;
+			w_flag.store(false,std::memory_order_release);
+
 			double t_start = MPI_Wtime();
-			std::lock_guard<std::mutex> lock(_buffer_mutex);
+			// Delegate flushing to lock holder (can be self)
+			_qd_lock.DELEGATE_N(&write_buffer::_flush_helper, this, &w_flag);
+			double t_end = MPI_Wtime();
+
+			// Wait until flush is completed
+			while(!w_flag.load(std::memory_order_acquire));
```

Comment:
is there, or should there be a statistic for how much we burn CPU time here? High amounts of time spent in this line could indicate we are preventing other threads from doing useful work.